### PR TITLE
Refactoring of Storage Adapters

### PR DESCRIPTION
**Pull request by leventov**
- Factor `QueryableIndexColumnSelectorFactory` and `IncrementalIndexColumnSelectorFactory` out of `QueryableIndexStorageAdapter` and `IncrementalIndexStorageAdapter` respectively. This change is the main purpose of this PR: those classes are going to be reused later in #4622.
 - Reduce indirection in `ColumnValueSelector`s created by `QueryableIndexColumnSelectorFactory`, partially thanks to static `Offset`s now after #4706.
 - Remove `OffsetHolder` interface.
 - Add `Offset.getBaseReadableOffset()`
 - Instead of `Cursor extends ColumnSelectorFactory`, add `Cursor.getColumnSelectorFactory()`, i. e. replace inheritance with composition.
 - Rename `IncrementalIndexStorageAdapter.EntryHolder` to `TimeAndDimsHolder`

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java [hunk 1]
```diff
@@ -435,487 +419,58 @@ public Cursor apply(final Interval inputInterval)
                                         );
 
 
-                  final Offset initOffset = offset.clone();
+                  final Offset cursorOffset = offset.clone();
+                  QueryableIndexColumnSelectorFactory columnSelectorFactory = new
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: to fit convention in here, please add `final`

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java [hunk 2]
```diff
@@ -435,487 +419,58 @@ public Cursor apply(final Interval inputInterval)
                                         );
 
 
-                  final Offset initOffset = offset.clone();
+                  final Offset cursorOffset = offset.clone();
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: can this be changed back to `initOffset` (or something more descriptive)? it is really hard to keep track of if `this.cursorOffset` or `cursorOffset` is the correct usage in this method.

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java (hunk 1 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: Fixed

File: processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java]: I simplified this area a bit and removed anonymous subclasses. Also renamed the variable to `baseCursorOffset`. "initOffset" is not quite right because in the previous version of the code, "initOffset" was immutable and always pointing to the "init offset" (and cloned). Now this "cursorOffset" is actually mutable and resettable inside the cursor.

File: processing/src/main/java/io/druid/segment/column/GenericColumn.java [hunk 3]
```diff
@@ -39,16 +43,24 @@
   public String getStringSingleValueRow(int rowNum);
   @CalledFromHotLoop
   public Indexed<String> getStringMultiValueRow(int rowNum);
+
   @CalledFromHotLoop
-  public float getFloatSingleValueRow(int rowNum);
+  float getFloatSingleValueRow(int rowNum);
+  HistoricalFloatColumnSelector makeFloatSingleValueRowSelector(ReadableOffset offset);
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/column/GenericColumn.java]: Should be FloatColumnSelector? or better to change the return types of the below methods because this class is used for only QueryableIndex?

File: processing/src/main/java/io/druid/segment/column/IndexedDoublesGenericColumn.java [hunk 4]
```diff
@@ -72,6 +76,12 @@ public float getFloatSingleValueRow(int rowNum)
   }
 
   @Override
+  public HistoricalFloatColumnSelector makeFloatSingleValueRowSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/column/IndexedDoublesGenericColumn.java]: Same here.

File: processing/src/main/java/io/druid/segment/column/IndexedFloatsGenericColumn.java [hunk 5]
```diff
@@ -72,6 +76,12 @@ public float getFloatSingleValueRow(int rowNum)
   }
 
   @Override
+  public HistoricalFloatColumnSelector makeFloatSingleValueRowSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/column/IndexedFloatsGenericColumn.java]: same here.

File: processing/src/main/java/io/druid/segment/column/IndexedLongsGenericColumn.java [hunk 6]
```diff
@@ -72,6 +76,12 @@ public float getFloatSingleValueRow(int rowNum)
   }
 
   @Override
+  public HistoricalFloatColumnSelector makeFloatSingleValueRowSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/column/IndexedLongsGenericColumn.java]: same here.

File: processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java [hunk 7]
```diff
@@ -105,10 +105,7 @@ public int getCardinality()
   }
 
   @Override
-  public HistoricalDimensionSelector makeDimensionSelector(
-      final OffsetHolder offsetHolder,
-      final ExtractionFn extractionFn
-  )
+  public HistoricalDimensionSelector makeDimensionSelector(final ReadableOffset offset, final ExtractionFn extractionFn)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java]: same here.

File: processing/src/main/java/io/druid/segment/data/IndexedDoubles.java [hunk 8]
```diff
@@ -29,5 +34,68 @@
...
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        inspector.visit("indexed", IndexedDoubles.this);
+        inspector.visit("offset", offset);
+      }
+    };
+  }
+
+  default HistoricalFloatColumnSelector makeFloatColumnSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/data/IndexedDoubles.java]: same here.

File: processing/src/main/java/io/druid/segment/data/IndexedFloats.java [hunk 9]
```diff
@@ -32,4 +37,67 @@
 
   @Override
   void close();
+
+  default HistoricalFloatColumnSelector makeFloatColumnSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/data/IndexedFloats.java]: same here.

File: processing/src/main/java/io/druid/segment/data/IndexedLongs.java [hunk 10]
```diff
@@ -32,4 +37,67 @@
...
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        inspector.visit("indexed", IndexedLongs.this);
+        inspector.visit("offset", offset);
+      }
+    };
+  }
+
+  default HistoricalFloatColumnSelector makeFloatColumnSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/data/IndexedLongs.java]: same here.

File: processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java [hunk 11]
```diff
@@ -17,19 +17,33 @@
...
+  }
+
+  public void set(IncrementalIndex.TimeAndDims currEntry)
+  {
+    this.currEntry = currEntry;
+  }
+
+  public IncrementalIndex.TimeAndDims getKey()
+  {
+    return currEntry;
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java]: Is this method supposed to return ```currEntry``` as ```get()``` does? Also, what do the key and value mean here?

Comment (leventov) [reply processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java]: I agree that this class has strange API. In the future PRs in this series I will probably reduce it to just get() + set(). However in this PR I just renamed and moved this class to upper level, without other modifications.

File: processing/src/main/java/io/druid/segment/column/GenericColumn.java (hunk 3 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/column/GenericColumn.java]: I didn't understand this. Why return types should be changed?

Comment (jihoonson) [reply processing/src/main/java/io/druid/segment/column/GenericColumn.java]: For consistency. makeLongSingleValueRowSelector() and makeDoubleSingleValueRowSelector() return LongColumnSelector and DoubleColumnSelector, respectively.

Comment (jihoonson) [reply processing/src/main/java/io/druid/segment/column/GenericColumn.java]: Also, please remove ```getXxxxMultiValueRow()``` methods because they are not used anymore.

File: processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java (hunk 11 above)
Comment (jihoonson) [reply processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java]: Ok, then please leave a comment about this.

Comment (leventov) [reply processing/src/main/java/io/druid/segment/incremental/TimeAndDimsHolder.java]: Added comments

File: processing/src/main/java/io/druid/segment/column/GenericColumn.java (hunk 3 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/column/GenericColumn.java]: Interfaces like `HistoricalLongColumnSelector` and `HistoricalDoubleColumnSelector` don't exist yet. When they are added, the corresponding methods in GenericColumn and down the stack should be aligned with the current methods for the `float` specialization.

Comment (leventov) [reply processing/src/main/java/io/druid/segment/column/GenericColumn.java]: Removed `getXxxxMultiValueRow()`

**Comment by leventov** [comment]
@drcrallen do you have extra comments?

File: processing/src/main/java/io/druid/segment/FilteredOffset.java [hunk 12]
```diff
@@ -113,14 +88,32 @@ public boolean withinBounds()
...
+    return baseOffset.getBaseReadableOffset();
   }
 
   @Override
   public Offset clone()
   {
-    FilteredOffset offset = (FilteredOffset) super.clone();
-    offset.baseOffset = offset.baseOffset.clone();
-    return offset;
+    throw new UnsupportedOperationException("FilteredOffset should not be cloned");
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/segment/FilteredOffset.java]: @leventov why is this changed to throw an exception? More precisely, why are offsets cloned in topN?

Comment (leventov) [reply processing/src/main/java/io/druid/segment/FilteredOffset.java]: This code was never called and in fact it was wrong, because `filterMatcher` was not cloned. 

Offset.clone() is used for two purposes:
 1) making "frames" in QueryableIndexStorageAdapter
 2) (in theory) it helps JVM to figure out that the object could be scalarized on stack, within XxxPrototype.scanAndAggregate(). This method is inlined and JVM analyzes the scope of this method. If offset objects comes as parameter, it couldn't be scalarized. If it is cloned in the body of the method, it probably could be.

I don't remember was I actually successful to make JVM to scalarize this offset object or not, through.

Comment (gianm) [reply processing/src/main/java/io/druid/segment/FilteredOffset.java]: This causes some queries to fail-- raised an issue https://github.com/druid-io/druid/issues/5132.
